sstables/sstable_set.cc

```diff
@@ -769,6 +770,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+            _current_batch.push_back(std::move(mf));
+        } while (!_peeked_readers.empty()
+                    && _cmp(_current_batch.back().position(), _peeked_readers.front()->reader.peek_buffer().position()) == 0);
+
+        if (_unpeeked_readers.size() == 1 && _unpeeked_readers.front() == _galloping_reader) {
+            // The first condition says that only one reader was moved from the heap,
+            // i.e. all other readers had strictly greater positions.
+            // The second condition says that this reader already was a galloping candidate,
+            // so let's increase his score.
+            ++_gallop_mode_hits;
+
+            if (in_gallop_mode()) {
+                // We've entered gallop mode with _galloping_reader.
+                // In the next operator() call we will peek this reader on a separate codepath,
+                // using _galloping_reader instead of _unpeeked_readers.
+                _unpeeked_readers.clear();
+            }
+        } else {
+            // Each reader currently in _unpeeked_readers is a potential galloping candidate
+            // (they won with all other readers in _peeked_readers). Remember one of them.
+            _galloping_reader = _unpeeked_readers.front();
+            _gallop_mode_hits = 1;
+        }
+
+
+        collect_exhausted_readers(_current_batch.back().position());
+
+        return make_ready_future<mutation_fragment_batch>(_current_batch);
+    }
+
+    void next_partition() {
+        on_internal_error(cormlogger, "next_partition: this reader works only for single partition queries");
+    }
+
+
+    future<> fast_forward_to(const dht::partition_range&, db::timeout_clock::time_point) {
+        on_internal_error(cormlogger, "fast_forward_to: this reader works only for single partition queries");
+    }
+
+
+    future<> fast_forward_to(position_range pr, db::timeout_clock::time_point timeout) {
+        // Every reader in `_all_readers` has been peeked at least once, so it returned a partition_start.
+        // Thus every opened reader is safe to be fast forwarded.
+        _unpeeked_readers.clear();
+        _peeked_readers.clear();
+        _halted_readers.clear();
+        _galloping_reader = {};
+        _gallop_mode_hits = 0;
+
+        _unpeeked_readers.reserve(_all_readers.size());
+        for (auto it = _all_readers.begin(); it != _all_readers.end(); ++it) {
+            _unpeeked_readers.push_back(it);
+        }
+
+        _pr_end = pr.end();
+        _forwarded_to = pr;
+
+        return parallel_for_each(_unpeeked_readers, [this, pr = std::move(pr), timeout] (reader_iterator it) {
+            return it->reader.fast_forward_to(pr, timeout);
+        });
+    }
+
+    size_t buffer_size() const {
+        return std::accumulate(_all_readers.begin(), _all_readers.end(), size_t(0),
+                [] (size_t v, const reader_and_positions& r) { return v + r.reader.buffer_size(); });
+    }
```

Comment:
This method is now non-virtual and has an implementation in `flat_mutation_reader::impl`, no need for this.